--- subversion/libsvn_subr/sqlite.c.orig	2010-06-17 10:54:19.000000000 -0700
+++ subversion/libsvn_subr/sqlite.c	2012-05-15 15:21:31.000000000 -0700
@@ -740,6 +740,8 @@
 static svn_error_t *
 init_sqlite(void *baton, apr_pool_t *pool)
 {
+#if !__APPLE__
+// <rdar://problem/11438447> Xcode 4.4 subversion won't run on Lion
   if (sqlite3_libversion_number() < SQLITE_VERSION_NUMBER)
     {
       return svn_error_createf(
@@ -747,6 +749,7 @@
                     _("SQLite compiled for %s, but running with %s"),
                     SQLITE_VERSION, sqlite3_libversion());
     }
+#endif
 
 #if APR_HAS_THREADS
 
